snap/snapcraft.yaml

```diff
@@ -0,0 +1,93 @@
+name: micro-xrce-dds-agent
+base: core20
+version: git
+summary: Bridge between Micro XRCE-DDS clients and DDS
+description: |
+  Micro XRCE-DDS Agents receive messages containing Operations from
+  Clients. Agents keep track of the Clients and the Micro XRCE-DDS
+  Entities they create. The Agent uses the Entities to interact with
+  DDS Global Data Space on behalf of the Client.
+
+  The communication between a Micro XRCE-DDS Client and a Micro
+  XRCE-DDS Agent supports two kind transports: UDP or SerialPort.
+  While running Micro XRCE-DDS Agent will attend any received
+  request from your Micro XRCE-DDS Clients. Micro XRCE-DDS Agent
+  answers back with the result of a request each time a request is
+  attended.
+
+  Upon install, the Agent will be running as a daemon using `udp4`
+  on port 4000. The following parameters can be changed:
+
+  * `system`. Supported systems are `udp4`, `udp6`, `tcp4`, `tcp6`,
+    `serial`, and `pseudo-serial`. Default is `udp4`. Change with:
+
+        $ snap set micro-xrce-dds-agent system="new system"
+
+  * `middleware`. Supported kinds of middleware are `ced`, `rtps`, and
+    `dds`. Default is `dds`. Change with:
+
+        $ snap set micro-xrce-dds-agent middleware="new middleware"
+
+  * `verbosity`. Supported verbosity levels are 0-6, defaulting to 4.
+    Change with:
+
+        $ snap set micro-xrce-dds-agent verbosity="selected verbosity"
+
+  * `discovery`. Enable or disable the discovery server. Defaults to
+    "false". Change with:
+
+        $ snap set micro-xrce-dds-agent discovery="true or false"
+
+  * `discovery-port`. Port on which the discovery server (see above)
+    listens. Defaults to 7400. Change with:
+
+        $ snap set micro-xrce-dds-agent discovery-port="selected port"
+
+  * `p2p-port`. Port to use for the P2P profile. Change with:
+
+        $ snap set micro-xrce-dds-agent p2p-port="selected port"
+
+  * `port`. Port on which the agent listens. Only applicable to one of
+    the UDP or TCP systems (see above). Defaults to 4000. Change with:
+
+        $ snap set micro-xrce-dds-agent port="selected port"
+
+  * `baudrate`. Baud rate to use when accessing serial ports. Only
+    applicable when using the `serial` or `pseudo-serial` system.
+    Defaults to 115200. Change with:
+
+        $ snap set micro-xrce-dds-agent baudrate="baud rate"
+
+  * `device`. The serial device to use. Only applicable when using the
+    `serial` or `pseudo-serial` system. Change with:
+
+        $ snap set micro-xrce-dds-agent device="device path"
+
+
+grade: stable
+confinement: strict
+
+architectures:
+  - build-on: amd64
+  - build-on: arm64
+  - build-on: armhf
+  - build-on: ppc64el
+
+parts:
+  agent:
+    plugin: cmake
+    source: .
+    cmake-parameters: [-DCMAKE_INSTALL_PREFIX=/usr]
+    build-packages: [make, gcc, g++]
+
+  runner:
+    plugin: dump
+    source: snap/local/
+    organize:
+      '*': usr/bin/
+
+apps:
+  daemon:
+    command: usr/bin/snap-runner
+    daemon: simple
```

Comment:
A point of discussion: this line makes the agent a daemon, which makes the snap immediately useful upon install, but also means we need to settle on proper default behavior. Right now it's using `udp4` on port 4000, but that's not based on much real-world knowledge. Also, note that one can view the logs with `journalctl -f -u snap.micro-xrce-dds-agent.daemon`.

Reply:
Sounds good. There are a few ways to accomplish this;

First, we can go the other direction and get rid of the daemon altogether. If we expose the agent as a command that the user runs instead of a daemon and they want a daemon, they can turn it into one pretty trivially by writing a systemd unit for it (or whatever their init system is). This also allows us to get rid of the snapctl integration and just let the user hand CLI args to the agent like they always have.

On the other hand, if you like the way the daemon works (with the snapctl bits) we can default to having it disabled and document how to enable it. Might be some effort to keep the snapctl bits inline with the actual CLI interface if they aren't exercised regularly, so we'll need to keep an eye on that.

What is your favorite path?

Reply:
I've thought about this a little more and realized that the latter path is probably better-- it makes it possible to create an Ubuntu Core image with the daemon running on first boot whereas the former would make that type of thing harder.

Reply:
Sorry, yesterday I missed answering this one. We were discussing the two options and, for us, the second one was better. Having the `daemon` option indeed is a valuable asset, so that the user can switch on this behaviour once the developing stage is finished.
Also, having the possibility of launching the agent directly is exactly what we were looking for. I will try this throughout the morning and report back. Thank you!

Reply:
Oh no problem, sounds like we're on the same page here. Note that the snap is up-to-date with the latest commit here, so please do let me know what you think:

```
$ snap info micro-xrce-dds-agent
<snip>
channels:
  latest/stable:    v1.0.2+git107.51428f4 2020-09-02 (14) 4MB -
```